Validations require input via kwargs

gcheadle-vmware opened the pull request:
Contributes to #642 

## Feature Work

- [x] (@vmunishwar) allow assertion functions to be intuitively combined in boolean expressions — assertion functions return `True`, not `None`
- [x] (@pivotaljohn) support concise use of assertion functions in custom rules (e.g. `("custom def message", assert.min(1))` instead of `("custom def message", assert.min(1).check)`) — allow not just Callables, but also objects that have an `check` attribute that is a Callable.
- [x] implement `assert.not_null(v)` short-hand.
- [x] (@vmunishwar) store `min_len` and `max_len` as `starlark.Int`s (details: https://github.com/vmware-tanzu/carvel-ytt/pull/677#discussion_r910447266)
  - [x] remove comment (https://github.com/vmware-tanzu/carvel-ytt/pull/677#discussion_r910448304)
- [x] put new `@ytt:assert` functions into "validation" experiment to preserve the "right" to change the API.
  - [x] update docs for each such function with the same warning we have for `@schema/validation`
- [x] verify that duplicate validation kwargs are detected and result in an error.
- [x] switch the order of `fail()` and `True` for existing rules — to make the code more intuitive/readable.
- [x] `one_not_null()`/`one_not_null=`
  - [x] accept sequence of keys or bool for `assert.one_not_null()` and `one_not_null=`
  - [x] schema specific tests (`one_not_null=`)
  - [x] report helpful error message when value(s) is/are not a map.
- [x] Update GoDoc for `Max()`/`Min()`/`MaxLength()`/`MinLength()`
- [x] exploratory test edge-cases for inputs to validation kwargs.
  - when two validations rules contradict (e.g. min/max where max < min; min_len/max_len where max_len < min_len; etc.)
  - 0 cases
  - assert.<assertion-function>() with a misspelled name
  - where min=/max= are given a value that has no minimum or maximum defined, does the error message make sense? (e.g. None?)

## Fit and Finish

- [x] are there "TODO"s left in the code? Let's implement or remove.

gcheadle-vmware committed:
add min, max, min_len, max_len, not_null rules

- account for "not_null=" setting when determining if validations should
  be skipped.

Signed-off-by: Varsha Munishwar <vmunishwar@vmware.com>

pivotaljohn commented on pkg/validations/validate.go:
```diff
@@ -131,8 +132,10 @@ func (v *NodeValidation) DefaultNullSkipTrue() {
 func (v validationKwargs) shouldValidate(value starlark.Value, thread *starlark.Thread) (bool, error) {
-	_, isNone := value.(starlark.NoneType)
-	if isNone && (v.whenNullSkip != nil && *v.whenNullSkip) {
+	_, isNull := value.(starlark.NoneType)
+	nullAllowed := !v.notNull
+	whenNullSkip := v.whenNullSkip != nil && *v.whenNullSkip
+	if isNull && nullAllowed && whenNullSkip {
```
To keep this `if` from getting out of hand, captured each intention in a variable.

jtigger committed:
Make @ytt:assert filetests more navigable

... by splitting on function name.

Signed-off-by: Varsha Munishwar <vmunishwar@vmware.com>

jtigger committed:
Make yamlfragment comparable to their Starlark equivalent

... by yaml encode/decode'ing values for min() and max().

Signed-off-by: Varsha Munishwar <vmunishwar@vmware.com>

jtigger committed:
Introduce "assertion object" with check() attribute

- Simplifies the API, from:

  ```
  min(4)(5)
  min(4, 5)
  ```

  to a single, much more readable form:

  ```
  min(4).check(5)
  ```

- This new form also allows for potentially multiple values or either
  the assertion construction (i.e. the first call) or the check.
- introduces "Assertion" type
  - to ease use of such objects within Go and establish a concrete type
    within Starlark.
- increase test coverage over assert.max(), assert.min()
- makes boolean expressions more readable
- addresses https://github.com/vmware-tanzu/carvel-ytt/pull/677#discussion_r904349079

Co-authored-by: Varsha Munishwar <vmunishwar@vmware.com>

pivotaljohn commented on pkg/yttlibrary/assert.go:
```diff
@@ -261,21 +265,14 @@ func (b assertModule) Max(thread *starlark.Thread, f *starlark.Builtin, args sta
-		return starlark.None, err
-	}
-	return result, nil
+	return NewAssertNotNull(), nil
```
`assert.not_null()` is special: it does not take any parameters, so we just hand back the "assertion object", here. 👍🏻

gcheadle-vmware commented on pkg/validations/assert.go:
```diff
@@ -110,6 +119,31 @@ func newValidationKwargs(kwargs []starlark.Tuple, annPos *filepos.Position) (val
 			b := bool(v)
 			processedKwargs.whenNullSkip = &b
+		case ValidationKwargMinLength:
+			v, err := starlark.NumberToInt(value[1])
```
Should we document the rounding behavior of floats? or just document it as an int?

pivotaljohn replied:
Documenting as an integer, but accepting any number seems legit.

Starlark is dynamically typed; programmers will expect automatic coercion of numbers (maybe not necessarily 1.5 => 1, but 1.0 => 1, for sure).

gcheadle-vmware commented on pkg/validations/assert.go:
```diff
@@ -110,6 +119,31 @@ func newValidationKwargs(kwargs []starlark.Tuple, annPos *filepos.Position) (val
 			}
 			b := bool(v)
 			processedKwargs.whenNullSkip = &b
+		case ValidationKwargMinLength:
```
If we consider someone providing multiple `min_len` kwargs, would it make sense to take the greater minimum value since they are comparable numbers?

However, it is consistent if we always took the last one, which is how all these kwargs are implemented.

pivotaljohn replied:
I'm trying to come up with a use-case where an author _intentionally_ uses a keyword argument more than once...

... and the most likely scenario I can come up with is that they did it on accident ... which — if true — I bet they'd appreciate hearing about that.

So, you've got me starting to rethink the current behavior... that we'd rather error out on a duplicate keyword. 🤔 

Thoughts?

gcheadle-vmware commented on pkg/validations/validate.go:
```diff
@@ -30,6 +32,12 @@ type rule struct {
+	min          starlark.Value
+	max          starlark.Value
+	notNull      bool
+	// *int , default=nil possible_values={&1, &2, &-3, ..}
```
should this comment be on line 36?

pivotaljohn replied:
yeah! wooops!

pivotaljohn replied:
Okay... conversation below about storing these values as `starlark.Int` obviates this comment all together. Let's delete.

vmunishwar replied:
Done. Good catch, Garrett!

gcheadle-vmware commented on pkg/validations/validate.go:
```diff
@@ -125,8 +133,10 @@ func (v *NodeValidation) DefaultNullSkipTrue() {
-	_, isNone := value.(starlark.NoneType)
-	if isNone && (v.whenNullSkip != nil && *v.whenNullSkip) {
+	_, isNull := value.(starlark.NoneType)
+	nullAllowed := !v.notNull
```
I have to remind myself that `v.notNull` is false by default in order to understand the if statement below.

pivotaljohn replied:
excellent feedback. I see... `nullAllowed` has no apparent connection to the validation...  yeah, while there's a double negative (which is a little harder to reason) it's better than a full-on level of indirection. 👍🏻 

How about this?

```go
func (v validationKwargs) shouldValidate(value starlark.Value, thread *starlark.Thread) (bool, error) {
	_, valueIsNull := value.(starlark.NoneType)
	whenNullSkip := v.whenNullSkip != nil && *v.whenNullSkip
	if !v.notNull && whenNullSkip && valueIsNull {
		return false, nil
	}
```

gcheadle-vmware commented on pkg/validations/validate.go:
```diff
@@ -137,12 +147,53 @@ func (v validationKwargs) shouldValidate(value starlark.Value, thread *starlark.
+func (v validationKwargs) convertToRules() []rule {
+	var rules []rule
+	if minLen := v.minLength; minLen > 0 {
+		a := yttlibrary.NewAssertMinLen(core.NewGoValue(minLen).AsStarlarkValue())
```
I'm wondering if `minLen` and `maxLen` could just remain as their starlark types? It looks like we are converting it from a `starlark.Int` to an `int` when we construct `validationKwargs`, and then back to a starlark value here. We could save some sweat if we kept it as a starlark value the whole time.

The only concern I have with making such a change is that we are using the int in multiple places around this line of code (`minLen > 0`, `fmt.Sprintf("length greater or equal to %v", minLen)`).

pivotaljohn replied:
It's a good question.

A big driver for `min` and `max` to become `starlark.Value`s is that many (but not all) types in Starlark can be compared: we needed the wider type, there.
`min_len` and `max_len` _only_ max sense as an integer.

But to your point, we _could_ store these values as `starlark.Int`s... which have a convenient accessor to the underlying int64 value (for the usages you pointed at).

Let's do that!

gcheadle-vmware reviewed (commented):
Overall, this PR is shaping up really nicely. Really proud of our group effort! @vmunishwar, @pivotaljohn 
I've left my additional thoughts in the comments.

gcheadle-vmware commented on pkg/validations/assert.go:
```diff
@@ -59,8 +64,8 @@ func (a *convertAssertAnnsToValidations) Visit(node yamlmeta.Node) error {
 	var rules []rule
-
-	if len(annotation.Args) == 0 {
+	// TODO: what makes a malformed annotation??
```
Is this a malformed annotation? `#@schema/validation max_len=3, min_len=5`

pivotaljohn replied:
I'd say it's well-formed because it's syntactically correct.

However, it just won't work in practice. I have an assumption that library authors will test their business... and will discover that there's no set of data values that can satisfy all these rules.

gcheadle-vmware committed:
Add one_not_null to assert and validation kwargs

- combine main test cases into a single file to make it easier to see
  all the spec'ed behavior in one place.
- implement function as a Go func given how many moving parts there are.
- when one_not_null key is not present suggest nearest

Co-authored-by: John Ryan <jtigger@infosysengr.com>

vmunishwar committed:
assertion functions return True, not None

Signed-off-by: John Ryan <jtigger@infosysengr.com>

jtigger committed:
Provide assert.not_null(v) shorthand

- especially for cases outside of validation rules, the standard syntax
  is awkward/verbose: `assert.not_null().check(v)`

Co-authored-by: Garrett Cheadle <gcheadle@vmware.com>
Signed-off-by: Varsha Munishwar <vmunishwar@vmware.com>

jtigger committed:
Flag new @ytt:assert functions as "experimental"

Signed-off-by: Varsha Munishwar <vmunishwar@vmware.com>

vmunishwar committed:
store min_len and max_len as *starlark.Int

Co-authored-by: John Ryan <jtigger@infosysengr.com>

vmunishwar commented:
To be done later - review tests for duplicate coverage; every test represents maintenance, let's make sure they are doing their work.

vmunishwar committed:
switch the order of fail() and True for existing rules

Signed-off-by: John Ryan <jtigger@infosysengr.com>

jtigger committed:
Update GoDoc on @ytt:assert functions

- removed TODO: all known cases of malformed annotations _are_ covered.
  👍

Signed-off-by: Varsha Munishwar <vmunishwar@vmware.com>

jtigger committed:
Store "when=" as an interface value

cppforlife commented on pkg/yttlibrary/assert.go:
```diff
@@ -8,26 +8,51 @@ import (
+	}
+	if experiments.IsValidationsEnabled() {
+		members = starlark.StringDict{
+			"equals":       starlark.NewBuiltin("assert.equals", core.ErrWrapper(m.Equals)),
+			"fail":         starlark.NewBuiltin("assert.fail", core.ErrWrapper(m.Fail)),
+			"try_to":       starlark.NewBuiltin("assert.try_to", core.ErrWrapper(m.TryTo)),
```
instead of overriding members entirely, we should just be able to set new keys (keeping baseline the same)

jtigger committed:
Extract FileTests harness

... so that file tests can be more easily written/included in other
packages in the ytt module.

cppforlife commented on pkg/validations/assert.go:
```diff
@@ -71,45 +76,109 @@ func NewValidationFromValidationAnnotation(annotation template.NodeAnnotation) (
+		assertion, ok := ruleTuple[1].(starlark.Callable)
 		if !ok {
-			return nil, fmt.Errorf("expected second item in the 2-tuple to be an assertion function, but was %s (at %s)", ruleTuple[1].Type(), annotation.Position.AsCompactString())
+			var err error
```
nit: unnecessary decl

pivotaljohn replied:
In fact, the companion variable, `assertion` has a wider scope so this `var` is at least necessary, if not undesirable.

The scope of `err`, here, is only within this `if`, so seems appropriate.

cppforlife commented on pkg/validations/assert.go:
```diff
@@ -71,45 +76,109 @@ func NewValidationFromValidationAnnotation(annotation template.NodeAnnotation) (
+			case starlark.Sequence:
+				processedKwargs.oneNotNull = v
+			default:
+				return validationKwargs{}, fmt.Errorf("expected True or a sequence of keys, but was a \"%s\"", value[1].Type())
```
nit: we've typically used single quotes for user controlled values.

pivotaljohn replied:
There were a couple other places where this slipped in as well; fixed those too.

cppforlife commented on pkg/validations/validate.go:
```diff
@@ -125,24 +133,90 @@ func (v *NodeValidation) DefaultNullSkipTrue() {
-	if isNone && (v.whenNullSkip != nil && *v.whenNullSkip) {
+	_, valueIsNull := value.(starlark.NoneType)
+	whenNullSkip := v.whenNullSkip != nil && *v.whenNullSkip
+	if !v.notNull && whenNullSkip && valueIsNull {
```
todo do we really need `!v.notNull` since valueIsNull==true

pivotaljohn replied:
Turns out: we do.

`!v.notNull` should be read, "is null allowed?" ... so we should only entertain skipping validating this value if null is allowed.

introduced a variable to improve the readability, here.

cppforlife commented on pkg/validations/validate.go:
```diff
@@ -125,24 +133,90 @@ func (v *NodeValidation) DefaultNullSkipTrue() {
 
-	if v.when != nil {
-		result, err := starlark.Call(thread, *v.when, starlark.Tuple{value}, []starlark.Tuple{})
+	if !(v.when == nil || reflect.ValueOf(v.when).IsNil()) {
```
nit: more readable with ! + &&

cppforlife commented on pkg/validations/validate.go:
```diff
@@ -125,24 +133,90 @@ func (v *NodeValidation) DefaultNullSkipTrue() {
 
 		isTrue := bool(result.Truth())
-		return isNone || isTrue, nil
+		return valueIsNull || isTrue, nil
```
valueIsNull? are we saying we should always validate for null values.

cppforlife commented on pkg/validations/validate.go:
```diff
@@ -125,24 +133,90 @@ func (v *NodeValidation) DefaultNullSkipTrue() {
+	var rules []rule
+
+	if v.minLength != nil {
+		a := yttlibrary.NewAssertMinLen(*v.minLength)
```
nit: inline assertion constuctor

pivotaljohn replied:
... and remaining `a` expanded to the more selectable, `assertion`. 👍🏻

cppforlife commented on pkg/yttlibrary/assert.go:
```diff
@@ -101,3 +129,289 @@ func (b assertModule) TryTo(thread *starlark.Thread, f *starlark.Builtin, args s
+
+// Assertion encapsulates a rule (a predicate) that can be accessed in a Starlark expression (via the "check" attribute)
+// or in Go (via CheckFunc()).
+type Assertion struct {
```
lets add ConversionHint

jtigger committed:
Move validations-specific tests to "validations" pkg

- using the filetest format, instead to improve readability.

jtigger committed:
Fixups per review feedback

- numerous renames to increase selectivity for easier editing across
  multiple lines
- other renames where simpler names are clearer
- ensure that any user-supplied value is surrounded by single quotes (')
  not double-quotes (") as is our convention.
- for syntax errors around the number of arguments, use the exact same
  format as Starlark does for more consistency.
- supply GoDoc for exported symbols to improve code readability
  (sometimes questionable, but we're living with those so as to
   encourage documentation in code).
- exercise `struct` values in the assertion functions where they make
  sense to ensure that users who want to do so are not surprised by an
  error or different behavior.

jtigger committed:
Reorg test coverage of validations

- generally:
  - test from within the package that is responsible for a given
    behavior
  - prefer filetests for single-file scenarios -- easier to maintain and
    read
- pkg/validations -- parsing and running validations
- cmd/template -- given validations in schema (and data values), ensure
  they combine as intended
